extensions/wikia/WallNotifications/WallNotifications.class.php

```diff
@@ -30,70 +30,75 @@ public function __construct() {
+	 * @param int $readSlice Max number of already read messages to show
+	 * @param bool $countOnly Set to true to only return a count.  This should be split out to another function
+	 * @param bool $notifyEveryone
+	 *
+	 * @return array|Mixed
+	 */
+	public function getWikiNotifications( $userId, $wikiId, $readSlice = 5, $countOnly = false, $notifyEveryone = false ) {
 
+		$list = $this->getWikiNotificationList( $userId, $wikiId );
 		if ( empty( $list ) ) {
-			wfProfileOut(__METHOD__);
-			return [];
+			return $list;
 		}
+
 		$read = [];
 		$unread = [];
+		$cacheCleared = false;
 
-		// walk through list of ids
-		// $listval is unique_id field from wall_notification table in DB
-		foreach ( array_reverse( $list['notification'] ) as $listval ) {
-			if ( !empty( $listval ) ) {
-				if ( !$countonly ) {
-					$grouped = $this->groupEntity( $list['relation'][ $listval ]['list'] );
-				} else {
-					$grouped = [];
+		// walk through list of ids from wall_notification table in DB
+		foreach ( array_reverse( $list['notification'] ) as $notifyUniqueId ) {
+			$notifyRelation = $list['relation'][$notifyUniqueId];
+
+			if ( empty( $notifyUniqueId ) ) {
+				continue;
+			}
+
+			$grouped = [];
+			if ( !$countOnly ) {
+				$grouped = $this->groupEntity( $notifyRelation['list'] );
+
+				if ( empty( $grouped ) ) {
+					continue;
 				}
+			}
 
-				if ( !empty( $grouped ) || $countonly ) {
-					if ( $list['relation'][ $listval ]['read'] ){
-						if ( count( $read ) < $readSlice ){
-							$read[] = [
-								"grouped" => $grouped,
-								"count" => 	empty( $list['relation'][ $listval ]['count'] )
-											? count( $list['relation'][ $listval ]['list'] )
-											: $list['relation'][ $listval ]['count']
-							];
-						} elseif ( $readSlice > 0 ) {
-							// so we have more read notifications that we need for display
-							// remove them
-							$this->remNotificationsForUniqueID( $userId, $wikiId, $listval );
-						}
-					} else {
-						if ( empty( $list['relation'][ $listval ]['notifyeveryone']) || $notifyeveryone ) {
-							$unread[] = [
-								"grouped" => $grouped,
-								"count" => 	empty( $list['relation'][ $listval ]['count'] )
-											? count( $list['relation'][ $listval ]['list'] )
-											: $list['relation'][ $listval ]['count']
-							];
-						}
+			if ( $notifyRelation['read'] ) {
+				if ( count( $read ) < $readSlice ) {
+					$read[] = [
+						"grouped" => $grouped,
+						"count" => 	empty( $notifyRelation['count'] )
+									? count( $notifyRelation['list'] )
+									: $notifyRelation['count']
+					];
+				} elseif ( $readSlice > 0 ) {
+					// so we have more read notifications that we need for display
+					// remove them
+					$this->remNotificationsForUniqueID( $userId, $wikiId, $notifyUniqueId );
+
+					// After removing the notifications above, make sure we don't still
+					// have them in this cache and thus try to remove them again. SOC-815
+					if ( !$cacheCleared ) {
```

Comment:
@garthwebb Can you explain us more thoroughly how this change is going to impact on this code?
We're removing notifications synchronously in a loop, so I'm not sure if it's possible to call delete for the same notification many times. Was that caused by concurrent requests that were removing notifications for that user?